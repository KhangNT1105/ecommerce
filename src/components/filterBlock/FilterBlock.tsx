import React, { useState, useCallback, useRef, useEffect } from 'react';
import Styles from './FilterBlock.module.scss';
import { useTranslation } from 'react-i18next';

interface FilterBlockProps {
  filterList: string[];
  onItemChoose?: (item?: string) => void;
  className?: string;
}

const FilterBlock: React.FC<FilterBlockProps> = ({ filterList, className, onItemChoose }) => {
  const [showFilterList, setShowFilterList] = useState(false);
  const { t } = useTranslation();
  const filterBlockRef = useRef<HTMLDivElement | null>(null);

  const toggleFilterList = (isShow: boolean) => () => {
    setShowFilterList(isShow);
  };

  const onChoose = (item: string) => () => {
    onItemChoose && onItemChoose(item);
    setShowFilterList(false);
  };

  const onOutSideClick = useCallback((event) => {
    if (filterBlockRef && filterBlockRef!.current && !filterBlockRef!.current.contains(event.target)) {
      setShowFilterList(false);
    }
  }, []);

  useEffect(() => {
    document.addEventListener('click', onOutSideClick);
    return () => document.removeEventListener('click', onOutSideClick);
  }, [onOutSideClick]);

  return (
    <div className={className} ref={filterBlockRef}>
      <div className={Styles['filter-block__button']} onClick={toggleFilterList(!showFilterList)}>
        <div>{t('ADD_FILTER')}</div>
        <div className="icon-add" />
      </div>

      {showFilterList && filterList.length > 0 && (
        <div className={Styles['filter-block__filter-list']}>
          <div>
            {filterList.map((filterElm) => (
              <div key={filterElm} onClick={onChoose(filterElm)}>
                {filterElm}
              </div>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default FilterBlock;
