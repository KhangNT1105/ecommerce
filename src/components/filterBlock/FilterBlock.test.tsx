import React from 'react';
import renderer from 'react-test-renderer';
import FilterBlock from './FilterBlock';
import { shallow, mount } from 'enzyme';
import ReactDOM from 'react-dom';

describe('component FilterBlock', () => {
  test('renders snapshot correctly', () => {
    const tree = renderer.create(<FilterBlock filterList={['id', 'name', 'this is a very long item']} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('FilterBlock event', () => {
  test('click on filter button', () => {
    const instance = shallow(<FilterBlock className="test" filterList={['id', 'name', 'this is a very long item']} />);
    instance.find('.test > div:first-child').first().simulate('click');
  });

  test('click on filter button', () => {
    const testFunc = jest.fn();

    const instance = shallow(
      <FilterBlock onItemChoose={testFunc} className="test" filterList={['id', 'name', 'this is a very long item']} />
    );
    instance.find('.test > div:first-child').first().simulate('click');
    instance.find('.test > div:last-child > div > div:first-child').first().simulate('click');

    expect(testFunc).toHaveBeenCalled();
  });

  test('outside click', () => {
    interface EventInterface {
      [name: string]: any;
    }

    const event: EventInterface = {};

    const addEventListener = jest.spyOn(document, 'addEventListener').mockImplementation((key, func) => {
      event[key] = func;
    });
    const instance = mount(<FilterBlock className="test" filterList={['id', 'name', 'this is a very long item']} />);

    instance.find('.test > div').first().simulate('click');

    event.click(<div />);
    expect(instance.find('.test > div:last-child > div > div').length).toEqual(0);
    addEventListener.mockRestore();
  });

  test('inside click', () => {
    interface EventInterface {
      [name: string]: any;
    }

    const event: EventInterface = {};

    const addEventListener = jest.spyOn(document, 'addEventListener').mockImplementation((key, func) => {
      event[key] = func;
    });

    const instance = mount(<FilterBlock className="test" filterList={['id', 'name', 'this is a very long item']} />);

    event.click({
      target: ReactDOM.findDOMNode(instance.find('.test').last().instance())
    });
    expect(instance.find('.test>div:last-child>div>div').length).toEqual(0);
    addEventListener.mockRestore();
  });

  test('unmount', () => {
    const removeEvent = jest.spyOn(document, 'removeEventListener');
    const instance = mount(<FilterBlock className="test" filterList={['id', 'name', 'this is a very long item']} />);
    instance.unmount();
    expect(removeEvent).toHaveBeenCalled();
  });
});
