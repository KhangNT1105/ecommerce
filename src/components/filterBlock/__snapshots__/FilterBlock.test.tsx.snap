// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component FilterBlock renders snapshot correctly 1`] = `
<div>
  <div
    className="filter-block__button"
    onClick={[Function]}
  >
    <div>
      ADD_FILTER
    </div>
    <div
      className="icon-add"
    />
  </div>
</div>
`;
